services/kafka/service.go

```diff
@@ -0,0 +1,284 @@
+		old.SSLCA != new.SSLCA ||
+		old.SSLCert != new.SSLCert ||
+		old.SSLKey != new.SSLKey
+}
+
+func (c *Cluster) clearWriters() {
+	for t, w := range c.writers {
+		w.Close()
+		delete(c.writers, t)
+	}
+}
+
+type Service struct {
+	mu       sync.RWMutex
+	clusters map[string]*Cluster
+	diag     Diagnostic
+}
+
+func NewService(cs Configs, d Diagnostic) *Service {
+	clusters := make(map[string]*Cluster, len(cs))
+	for _, c := range cs {
+		if c.InsecureSkipVerify {
+			d.InsecureSkipVerify()
+		}
+		clusters[c.ID] = NewCluster(c)
+	}
+	return &Service{
+		diag:     d,
+		clusters: clusters,
+	}
+}
+
+func (s *Service) Cluster(id string) (*Cluster, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	c, ok := s.clusters[id]
+	return c, ok
+}
+func (s *Service) Update(newConfigs []interface{}) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	clusterExists := make(map[string]bool, len(s.clusters))
+
+	for _, nc := range newConfigs {
+		if c, ok := nc.(Config); ok {
+			if err := c.Validate(); err != nil {
+				return err
+			}
+			if c.Enabled {
+				if c.InsecureSkipVerify {
+					s.diag.InsecureSkipVerify()
+				}
+				cluster, ok := s.clusters[c.ID]
+				if !ok {
+					s.clusters[c.ID] = NewCluster(c)
+				} else {
+					if err := cluster.Update(c); err != nil {
+						return errors.Wrapf(err, "failed to update cluster %q", c.ID)
+					}
+				}
+				clusterExists[c.ID] = true
+			} else {
+				cluster, ok := s.clusters[c.ID]
+				if ok {
+					cluster.Close()
+					delete(s.clusters, c.ID)
+				}
+			}
+		} else {
+			return fmt.Errorf("unexpected config object type, got %T exp %T", nc, c)
+		}
+	}
+
+	// Find any deleted clusters
+	for name, cluster := range s.clusters {
+		if !clusterExists[name] {
+			cluster.Close()
+			delete(s.clusters, name)
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) Open() error {
+	return nil
+}
+
+func (s *Service) Close() error {
+	return nil
+}
+
+type testOptions struct {
+	Cluster string `json:"cluster"`
+	Topic   string `json:"topic"`
+	Key     string `json:"key"`
+	Message string `json:"message"`
+}
+
+func (s *Service) TestOptions() interface{} {
+	return &testOptions{
+		Cluster: "example",
+		Topic:   "test",
+		Key:     "key",
+		Message: "test kafka message",
+	}
+}
+
+func (s *Service) Test(options interface{}) error {
+	o, ok := options.(*testOptions)
+	if !ok {
+		return fmt.Errorf("unexpected options type %t", options)
+	}
+	c, ok := s.Cluster(o.Cluster)
+	if !ok {
+		return fmt.Errorf("unknown cluster %q", o.Cluster)
+	}
+	return c.WriteMessage(o.Topic, []byte(o.Key), []byte(o.Message))
+}
+
+type HandlerConfig struct {
+	Cluster  string `mapstructure:"cluster"`
+	Topic    string `mapstructure:"topic"`
+	Template string `mapstructure:"template"`
+}
+
+type handler struct {
+	s *Service
+
+	cluster  *Cluster
+	topic    string
+	template *template.Template
+
+	diag Diagnostic
+}
+
+func (s *Service) Handler(c HandlerConfig, ctx ...keyvalue.T) (alert.Handler, error) {
+	cluster, ok := s.Cluster(c.Cluster)
```

Comment:
Do we want to allow for a default cluster/topic?  So they can do 
alert() 
  .kafka()  

as opposed to 
alert() 
  .kafka() 
  .cluster("default")
  .kafkaTopic("q1")

The slack update does that, but that's mostly to preserve backward compatibility with the old config format.  Probably better to just require them to be explicit

Reply:
At this point I'd prefer them to be explicit. It simplifies the code quite a bit. And there is no existing behavior to preserve.